### ISSUE TITLE
Composable indexes with multiple different configurations

### ISSUE THREAD

**Issue by fgiasson** [jerryjliu/llama_index#362]
I am trying to create a composable index with multiple Faiss indexes, but I have some issues getting the configuration right. 

Initialy I created a `GPTListIndex`with 3 `SimpleDict` which worked fine. The first config was used for the `List`and the second one for each of the `SimpleDict`. 

However, when I switched to `GPTFaissIndex` it stopped worked in the following way.

I create a `GPTListIndex` that way:

```python
list_index = GPTListIndex([faiss_1, faiss_2])
```

Then I create the config that way:

```python
query_configs = [
    QueryConfig(
        index_struct_type=IndexStructType.LIST,
        query_mode=QueryMode.DEFAULT,
        query_kwargs={}
    ),
    QueryConfig(
        index_struct_type=IndexStructType.DICT,
        query_mode=QueryMode.DEFAULT,
        query_kwargs={'faiss_index': faiss_1._faiss_index}
    ),
    QueryConfig(
        index_struct_type=IndexStructType.DICT,
        query_mode=QueryMode.DEFAULT,
        query_kwargs={'faiss_index': faiss_2._faiss_index}
    )
]
```

I was expecting that the `faiss_1` config would be used for the first faiss index in the list, and the second config for the second, etc. by the query runner. However, it doesn't seem to be the case. I am getting that errror:

```
ValueError: text_id not found in id_map
```

Which is because the `faiss_1.faiss._index` is used for faiss_2.

If I do the following, everything works fine for a single index:

```python
list_index = GPTListIndex([faiss_1])
```

Then I create the config that way:

```python
query_configs = [
    QueryConfig(
        index_struct_type=IndexStructType.LIST,
        query_mode=QueryMode.DEFAULT,
        query_kwargs={}
    ),
    QueryConfig(
        index_struct_type=IndexStructType.DICT,
        query_mode=QueryMode.DEFAULT,
        query_kwargs={'faiss_index': faiss_1._faiss_index}
    )
]
```

And if I rever 1 for 2, it works as expected as well.

I am not sure if there is something I don't understand with how the composability API works, or if there is a bug or missing feature somewhere.

I am using version `0.3.3`. I will continue to debug while waiting to see if someone has encountered this issue. I will update the doc if need be once I found the issue.

**Comment by fgiasson** [comment]
Note: I just noticed in the `0.3.3` codebase that the new preferred way is to use a `ComposableGraph`, but the same issue happen as far as I can see.

**Comment by fgiasson** [comment]
Some more, normally I don't think we are supposed to add the `._faiss_index` in the `query_kwargs` like above. However, if we don't, then a `faiss_index connot be None` [error from here](https://github.com/jerryjliu/gpt_index/blob/1fa174fab5465a5b267d049f5b3419c3d7997c24/gpt_index/indices/query/vector_store/faiss.py#L53).

I would have expected that this index get propagated somehow [like here](https://github.com/jerryjliu/gpt_index/blob/1fa174fab5465a5b267d049f5b3419c3d7997c24/gpt_index/indices/vector_store/faiss.py#L128), but it doesn't look like to be the case.

**Comment by jerryjliu** [comment]
Hi @fgiasson, so sorry for the delayed response. 

I think this may be a design bug in the composability API. For your reference, the way we define query configs is not keyed by the id of the index, but rather by the type of index. So we didn't really account for having two query configs of the same type as you listed here: 

```
query_configs = [
    QueryConfig(
        index_struct_type=IndexStructType.LIST,
        query_mode=QueryMode.DEFAULT,
        query_kwargs={}
    ),
    QueryConfig(
        index_struct_type=IndexStructType.DICT,
        query_mode=QueryMode.DEFAULT,
        query_kwargs={'faiss_index': faiss_1._faiss_index}
    ),
    QueryConfig(
        index_struct_type=IndexStructType.DICT,
        query_mode=QueryMode.DEFAULT,
        query_kwargs={'faiss_index': faiss_2._faiss_index}
    )
]
```

Of course, this doesn't work for the faiss index, which requires that different faiss indices are specified for each query call. Let me take a deeper look at this soon!

**Comment by fgiasson** [comment]
Hi @jerryjliu! 

Yeah no worries, everyone are busy 😄 

I thought it would be something like that, but I couldn't confirm since I didn't yet had time to debug line-by-like to better understand the design. I think it will become a requirement for the design. 

What I was doing is to use `gpt-index`to spawn a lot of different datasets, slice-and-diced differently, to combine them in higher order indexes to see the impact on the results. This was during the experimentation phase. And then, the best ones were kept and maybe consolidated in a single new one for faster execution.

**Comment by jerryjliu** [comment]
this was fixed in #572 , you can specify id when defining a query config
